lib/request_handler/validation/dry_engine.rb

```diff
@@ -7,34 +7,45 @@ module RequestHandler
   module Validation
     class DryEngine < Engine
       def self.valid_schema?(schema)
-        schema.respond_to?(:call)
+        schema_instance?(schema) || schema_class?(schema)
       end
 
       def self.validate(value, schema, options: {})
-        value = value.deep_symbolize_keys if schema.respond_to?(:rules) &&
-                                             schema.rules.keys.first.is_a?(Symbol)
-        result = if options.empty?
-                   schema.call(value)
-                 else
-                   schema.with(options).call(value)
-                 end
-
-        to_result(result)
+        value = value.deep_symbolize_keys if value.is_a?(Hash)
+        to_result(call_schema(value, schema, options))
       rescue Dry::Types::ConstraintError => e
         Result.new(output: nil, errors: { '' => e })
       end
 
+      def self.call_schema(value, schema, options)
+        if options.empty?
+          schema_instance?(schema) ? schema.call(value) : schema.new.call(value)
+        else
+          schema_instance?(schema) ? schema.with(options).call(value) : schema.new(options).call(value)
+        end
+      end
+
+      def self.schema_instance?(schema)
+        schema.respond_to?(:call)
+      end
+
+      def self.schema_class?(schema)
+        schema.respond_to?(:schema)
+      end
+
       def self.validate!(value, schema, options: {})
         validate(value, schema, options).tap do |result|
-          raise Validation::Error unless result.valid?
+          valid = result.respond_to?(:valid?) ? result.valid? : result.success?
+          raise Validation::Error unless valid
         end
       end
 
       def self.to_result(result)
-        if result.is_a?(Dry::Validation::Result)
-          Result.new(output: result.output, errors: result.errors)
+        if result.is_a?(Dry::Validation::Result) || result.is_a?(Dry::Schema::Result)
+          Result.new(output: result.to_h, errors: result.errors.to_h)
```

Comment:
can't you merge these 2 if paths? in both cases you do ` to_h` on the result and one has no error while for the other you just take `result.errors.to_h`

Reply:
Simplified it a little bit!